Kernel/ACPI/ACPIStaticParser.cpp

```diff
@@ -0,0 +1,376 @@
+}
+
+void ACPIStaticParser::init_fadt()
+{
+    kprintf("ACPI: Initializing Fixed ACPI data\n");
+    ACPI_RAW::SDTHeader& physical_fadt = find_fadt();
+
+    auto checkup_region = MM.allocate_kernel_region((PAGE_SIZE * 2), "ACPI Static Parser", Region::Access::Read);
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: Checking FADT Length to choose the correct mapping size\n");
+#endif
+    mmap_region(*checkup_region, PhysicalAddress((u32)(&physical_fadt) & PAGE_MASK));
+
+    ACPI_RAW::SDTHeader* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().get() + ((u32)(&physical_fadt) & (~PAGE_MASK)));
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: FADT @ V 0x%x, P 0x%x\n", sdt, &physical_fadt);
+#endif
+    u32 length = sdt->length;
+    kprintf("ACPI: Fixed ACPI data, Revision %u\n", sdt->revision);
+
+    auto fadt_region = MM.allocate_kernel_region(PAGE_ROUND_UP(length) + PAGE_SIZE, "ACPI Static Parser", Region::Access::Read);
+    mmap_region(*fadt_region, PhysicalAddress((u32)(&physical_fadt) & PAGE_MASK));
+    m_fadt = adopt(*new ACPI::FixedACPIData(*((ACPI_RAW::FADT*)(fadt_region->vaddr().get() + ((u32)(&physical_fadt) & (~PAGE_MASK))))));
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: Finished to initialize Fixed ACPI data\n");
+#endif
+}
+
+void ACPIStaticParser::do_acpi_reboot()
+{
+    // FIXME: Determine if we need to do MMIO/PCI/IO access to reboot, according to ACPI spec 6.2, Section 4.8.3.6
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: Rebooting, Probing FADT (P @ 0x%x)\n", m_fadt.ptr());
+#endif
+    if (m_fadt->m_revision >= 2) {
+        kprintf("ACPI: Reboot, Sending value 0%x to Port 0x%x\n", m_fadt->m_reset_value, m_fadt->m_reset_reg.address);
+        IO::out8(m_fadt->m_reset_reg.address, m_fadt->m_reset_value);
+    } else {
+        kprintf("ACPI: Reboot, Not supported!\n");
+    }
+
+    ASSERT_NOT_REACHED(); /// If rebooting didn't work, halt.
+}
+
+void ACPIStaticParser::do_acpi_shutdown()
+{
+    kprintf("ACPI: Shutdown is not supported with the current configuration, Abort!\n");
+    ASSERT_NOT_REACHED();
+}
+
+void ACPIStaticParser::initialize_main_system_description_table()
+{
+    auto checkup_region = MM.allocate_kernel_region((PAGE_SIZE * 2), "ACPI Static Parser Copying Method", Region::Access::Read);
+#ifdef ACPI_DEBUG
+    dbgprintf("ACPI: Checking Main SDT Length to choose the correct mapping size");
+#endif
+    mmap_region(*checkup_region, PhysicalAddress((u32)m_main_system_description_table & PAGE_MASK));
+
+    auto* sdt = (ACPI_RAW::SDTHeader*)(checkup_region->vaddr().get() + ((u32)m_main_system_description_table & (~PAGE_MASK)));
+    u32 length = sdt->length;
+    u8 revision = sdt->revision;
+
+    auto main_sdt_region = MM.allocate_kernel_region(PAGE_ROUND_UP(length) + PAGE_SIZE, "ACPI Static Parser Copying Method", Region::Access::Read);
+    mmap_region(*main_sdt_region, PhysicalAddress((u32)m_main_system_description_table & PAGE_MASK));
+
+    Vector<ACPI_RAW::SDTHeader*>* sdt_pointers = new Vector<ACPI_RAW::SDTHeader*>();
```

Comment:
Please put this `Vector` on the stack instead:
```cpp
Vector<ACPI_RAW::SDTHeader*> sdt_pointers;
...
m_main_sdt = adopt(*new ACPI::MainSystemDescriptionTable(move(sdt_pointers)));
```

Otherwise, nobody is responsible for deleting it.

Reply:
Right, will fix it :)